### ISSUE TITLE
Question: Is there a plan to implement "promise.join" method recently?

### ISSUE THREAD

**Issue by nneesshh** [sandstorm-io/capnproto#452]
Just like title.
promise.exclusiveJoin means "when any promise fulfilled or rejected",
So, Maybe promise.join (include join) method means "when all promises fulfilled or any one is rejected"?
Will the code look like a chain? Such as "promise.join(...).join(...) "

**Comment by kentonv** [comment]
Hi @nneesshh,

There is `kj::joinPromises()`, which takes `kj::Array<kj::Promise<T>>`, returns `kj::Promise<kj::Array<T>>`, and resolves when all promises inside have resolved. Currently if one promise throws, it will still wait  for the others. I guess we could have a `joinPromisesFailfast()` which immediately rejects if any promise in the array rejects. I'd also like to have a version that can take promises of varying type, maybe returning a promise for a tuple, but that's more complicated...

**Comment by nneesshh** [comment]
Thank you vecy much, @kentonv, "joinPromisesFailfast" is what I expected.
I will try to use kj::joinPromises before 0.7, it also can solve my requirements if I force my task won't fail for ever.

**Comment by nneesshh** [comment]
Hi, @kentonv, I use the kj::joinPromises and I can't comose an arry of kj::Promise<void>.
I write code like this:
               "auto p1 = ...;
 		auto p2 = ...;
		auto arr = kj::heapArray<kj::Promise<void>>({ kj::mv(p1), kj::mv(p2) }); -- err c2280
		return kj::joinPromises(kj::mv(arr));"

But the visual studio 2015 compiler says: 
Error	C2280	'kj::Promise<void>::Promise(const kj::Promise<void> &)': attempting to reference a deleted function...	d:\dd\support\capnproto-master\c++\src\kj\array.h	667	

It seems like kj::Promise can't be construct through kj::heapArray.
Can you show me the proper way to use joinPromises?

**Comment by kentonv** [comment]
Hi @nneesshh,

To create an array of values that don't have default constructors, you need to use `ArrayBuilder`.

    auto builder = kj::heapArrayBuilder<kj::Pomise<T>>(size);
    builder.add(somePromise);
    builder.add(anotherPromise);
    ...
    kj::Array<kj::Promise<T>> array = builder.finish();

**Comment by nneesshh** [comment]
yes @kentonv , I got it now. I just missed the comment of heapArrayBuilder, many thanks to you.

**Comment by kentonv** [comment]
So, #706 solves this... but I have struggled to get it to work across compilers. It segfaults GCC and deeply confuses MSVC. I'm going to hold off on trying to fix those for now, so pushing this back to 0.8.
